crates/sargon/src/profile/mfa/security_structures/automatic_shield_builder/automatic_shield_builder.rs

```diff
@@ -0,0 +1,916 @@
+
+        let mut factors_to_add = self
+            .remaining_available_factors
+            .iter()
+            .filter(|&f| match selector {
+                FactorSelector::Category(category) => f.category() == category,
+                FactorSelector::Kind(kind) => f.factor_source_kind() == kind,
+            })
+            .map(|f| f.id())
+            .collect::<IndexSet<_>>();
+
+        if let Some(quantity) = quantity_to_add.as_fixed() {
+            factors_to_add = factors_to_add
+                .into_iter()
+                .take(quantity)
+                .collect::<IndexSet<_>>();
+        }
+
+        let number_of_factors_added = factors_to_add.len();
+        if number_of_factors_added == 0 {
+            return None;
+        }
+
+        self.remaining_available_factors
+            .retain(|f| !factors_to_add.contains(&f.id()));
+
+        self.proto_matrix
+            .add_factors_for_role(target_role, factors_to_add);
+
+        Some(number_of_factors_added)
+    }
+
+    fn factors_for_role(&self, role: RoleKind) -> &IndexSet<FactorSourceID> {
+        self.proto_matrix.factors_for_role(role)
+    }
+
+    /// Returns `true` if any factor was assigned, `false` otherwise.
+    fn assign_factors_of_category(
+        &mut self,
+        to: RoleKind,
+        category: FactorSourceCategory,
+        quantity_to_add: Quantity,
+    ) -> bool {
+        match self.assign_factors_matching_selector(
+            to,
+            FactorSelector::Category(category),
+            quantity_to_add,
+        ) {
+            Some(0) | None => false,
+            Some(_) => true,
+        }
+    }
+
+    /// Returns `true` if any factor was assigned, `false` otherwise.
+    fn assign_factors_of_kind(
+        &mut self,
+        to: RoleKind,
+        kind: FactorSourceKind,
+        quantity_to_add: Quantity,
+    ) -> bool {
+        match self.assign_factors_matching_selector(
+            to,
+            FactorSelector::Kind(kind),
+            quantity_to_add,
+        ) {
+            Some(0) | None => false,
+            Some(_) => true,
+        }
+    }
+
+    fn count_factors_for_role(&self, role_kind: RoleKind) -> u8 {
+        self.factors_for_role(role_kind).len() as u8
+    }
+
+    /// Returns `true` if any factor was assigned, `false` otherwise.
+    fn assign_factors_of_category_to_recovery(
+        &mut self,
+        category: FactorSourceCategory,
+        quantity_to_add: Quantity,
+    ) -> bool {
+        self.assign_factors_of_category(Recovery, category, quantity_to_add)
+    }
+
+    /// Returns `true` if any factor was assigned, `false` otherwise.
+    fn assign_factors_of_category_to_confirmation(
+        &mut self,
+        category: FactorSourceCategory,
+        quantity_to_add: Quantity,
+    ) -> bool {
+        self.assign_factors_of_category(Confirmation, category, quantity_to_add)
+    }
+
+    fn assign_factor_of_category_to_role_while_meaningful_and_less_than_limit(
```

Comment:
@ymatto I've updated to `loop` adding Factors to `role` of `category` while `cound_factors_for_role(role)` is `< limit` and this is called twice in `assign_custodian_and_hardware_to_role_while_meaningful_and_less_than_limit` below, once for `Custodian` (nothing happens for now, since unsupported) and and once for `Hardware` 

which is then in turn called twice in `assign_custodian_and_hardware_to_non_primary_roles_while_less_than_limit_for_each_assignment` below, first for `Recovery` and then for `Confirmation`, 

which is then in turn called twice in "main function" `assign` below, once with `limit: 1` and then again with `limit: 2`.

I think this is what you want?